tests/foreman/cli/test_repository.py

```diff
@@ -75,68 +83,332 @@
-            options['product-id'] = self.product['id']
+def _validated_image_tags_count(repo):
+    """Wrapper around Repository.info(), that returns once
+    container-image-tags in repo is greater than 0.
+    Needed due to BZ#1664631 (container-image-tags is not populated
+    immediately after synchronization), which was CLOSED WONTFIX
+    """
+    wait_for(
+        lambda: int(_get_image_tags_count(repo=repo)['content-counts']['container-image-tags'])
+        > 0,
+        timeout=30,
+        delay=2,
+        logger=logging.getLogger('robottelo'),
+    )
+    return _get_image_tags_count(repo=repo)
 
-        return make_repository(options)
 
-    def _get_image_tags_count(self, repo=None):
-        repo_detail = Repository.info({'id': repo['id']})
-        return repo_detail
+def _org(request):
+    org = make_org()
 
-    def _validated_image_tags_count(self, repo=None):
-        """Wrapper around Repository.info(), that returns once
-        container-image-tags in repo is greater than 0.
-        Needed due to BZ#1664631 (container-image-tags is not populated
-        immediately after synchronization), which was CLOSED WONTFIX
-        """
-        wait_for(
-            lambda: int(
-                self._get_image_tags_count(repo=repo)['content-counts']['container-image-tags']
-            )
-            > 0,
-            timeout=30,
-            delay=2,
-            logger=self.logger,
-        )
-        return self._get_image_tags_count(repo=repo)
+    @request.addfinalizer
+    def _cleanup():
+        if Org.exists(search=('id', org['id'])):
+            org_cleanup(org['id'])
+
+    return org
+
+
+def _product(request, org):
+    product = make_product_wait({'organization-id': org['id']})
+
+    @request.addfinalizer
+    def _cleanup():
+        if Product.exists(options={'name': product['name'], 'organization-id': org['id']}):
+            product_cleanup(product['id'])
+
+    return product
+
+
+def _repository(request, repo_options, org, product):
+    options = repo_options.copy()
+    options['organization-id'] = org['id']
+    options['product-id'] = product['id']
+
+    repo = make_repository(options)
+
+    @request.addfinalizer
+    def _cleanup():
+        if Repository.exists(
+            options={'name': repo['name'], 'organization-id': options['organization-id']}
+        ):
+            repo_cleanup(repo['id'])
+
+    return repo
+
+
+def _gpg_key(request, org):
+    gpg_key = make_gpg_key(options={'organization-id': org['id']})
+
+    @request.addfinalizer
+    def _cleanup():
+        if GPGKey.exists(options={'name': gpg_key['name'], 'organization-id': org['id']}):
+            gpg_key_cleanup(gpg_key['id'])
+
+    return gpg_key
+
+
+@pytest.fixture
+def repo_options(request):
+    """Return the options that were passed as indirect parameters."""
+    yield getattr(request, 'param', {}).copy()
+
+
+@pytest.fixture
+def org(request):
+    yield _org(request)
```

Comment:
Because the cleanup is registered with `request` fixture, these fixtures can just `return`. Functionally equivalent, but yield implies generator behavior or additional statements.

Reply:
Done.